sql/binlog.cc

```diff
@@ -9146,12 +9156,127 @@ static int binlog_start_trans_and_stmt(THD *thd, Log_event *start_event) {
 
     Query_log_event qinfo(thd, query, qlen, is_transactional, false, true, 0,
                           true);
-    if (cache_data->write_event(&qinfo)) DBUG_RETURN(1);
+    if (cache_data->write_event(thd, &qinfo)) DBUG_RETURN(1);
   }
 
   DBUG_RETURN(0);
 }
 
+/**
+  This function generated meta data in JSON format as a comment in a rows query
+  event.
+  @see binlog_trx_meta_data
+  @return JSON if all good, null string otherwise
+*/
+std::string THD::gen_trx_metadata() {
+  DBUG_ENTER("THD::gen_trx_metadata");
+  DBUG_ASSERT(opt_binlog_trx_meta_data);
+
+  ptree pt;
+  std::ostringstream buf;
+
+  // case: read existing meta data received from the master
+  if (rli_slave && !rli_slave->trx_meta_data_json.empty()) {
+    std::istringstream is(rli_slave->trx_meta_data_json);
+    try {
+      read_json(is, pt);
```

Comment:
It would make sense to switch to 8.0-native rapidjson, but probably as a follow-up work, time permitting